File: safe-commenter.js
import { TelegramClient } from 'telegram';
import { StringSession } from 'telegram/sessions/index.js';
import input from 'input';
import { NewMessage } from 'telegram/events/index.js';
import OpenAI from 'openai';
import dotenv from 'dotenv';
import fs from 'fs';
import path from 'path';
import RateLimiter from './rate-limiter.js';

// Load environment variables
dotenv.config();

// Load AI API
const openai = new OpenAI({
  apiKey: process.env.OPENROUTER_API_KEY,
  baseURL: process.env.OPENROUTER_API_URL,
});

// Telegram credentials
const apiId = parseInt(process.env.TELEGRAM_API_ID);
const apiHash = process.env.TELEGRAM_API_HASH;
const phoneNumber = process.env.TELEGRAM_PHONE_NUMBER;
const targetChannel = process.env.TARGET_CHANNEL;
// Add password - consider adding this to your .env file instead
if (!process.env.TELEGRAM_PASSWORD) {
  throw new Error('TELEGRAM_PASSWORD is not set in the environment. Please add it to your .env file.');
}
const password = process.env.TELEGRAM_PASSWORD;

// Rate limiter settings (can be customized)
const MAX_COMMENTS_PER_HOUR = 8;  // Conservative limit to avoid detection
const MAX_COMMENTS_PER_DAY = 50;  // Conservative limit to avoid detection

// Initialize rate limiter
const rateLimiter = new RateLimiter({
  maxPerHour: MAX_COMMENTS_PER_HOUR,
  maxPerDay: MAX_COMMENTS_PER_DAY,
  verbose: true
});

// Session file for persistence
const SESSION_FILE = './telegram_session.json';
const TEMP_DIR = './temp';

// Ensure temp directory exists
if (!fs.existsSync(TEMP_DIR)) {
  fs.mkdirSync(TEMP_DIR);
}

/**
 * Generate a comment based on post content using OpenAI
 * @param {Object} message - The Telegram message object
 * @param {Array} mediaDescriptions - Descriptions of any media in the message
 * @returns {string} - Generated comment
 */
async function generateComment(message, mediaDescriptions = []) {
  try {
    // Start with the text content
    let content = message.text || "";
    
    // Add media descriptions if available
    if (mediaDescriptions.length > 0) {
      content += "\n\nMedia in post: " + mediaDescriptions.join(", ");
    }
    
    // If there's no content to comment on, use a generic message
    if (!content || content.trim() === "") {
      content = "[Post contains media without text]";
    }

    const completion = await openai.chat.completions.create({
      model: process.env.MODEL_NAME,
      messages: [{
          role: "system",
          content: "Ты — персонаж Саске Учиха из аниме \"Наруто\". Отвечай холодно, сдержанно, немного высокомерно и загадочно, как он. Используй стиль речи Саске: короткие, чёткие фразы с запахом внутренней борьбы и силы. Будь иногда саркастичным и остроумным, не прямолинейным."
        },
        {
          role: "user",
          content: `Ответь на это сообщение как Саске: "${content}"`
        }
      ],
    });

    return completion.choices[0].message.content;
  } catch (error) {
    console.error('Error generating comment with OpenAI:', error);
    return "That's interesting! Thanks for sharing."; // Fallback comment
  }
}

/**
 * Process media in a message and generate descriptions
 * @param {Object} client - Telegram client
 * @param {Object} message - Telegram message
 * @returns {Array} - Array of media descriptions
 */
async function processMedia(client, message) {
  const mediaDescriptions = [];

  // Check if the message has media
  if (message.media) {
    const mediaType = message.media.className;
    
    switch (mediaType) {
      case 'MessageMediaPhoto':
        mediaDescriptions.push("Photo");
        break;
        
      case 'MessageMediaDocument':
        const doc = message.media.document;
        const attributes = doc.attributes;
        
        // Check for file name
        const fileNameAttr = attributes.find(attr => attr.className === 'DocumentAttributeFilename');
        const fileName = fileNameAttr ? fileNameAttr.fileName : 'document';
        
        // Check for video
        const videoAttr = attributes.find(attr => attr.className === 'DocumentAttributeVideo');
        if (videoAttr) {
          mediaDescriptions.push(`Video: ${fileName}`);
        } else {
          mediaDescriptions.push(`Document: ${fileName}`);
        }
        break;
        
      case 'MessageMediaPoll':
        const pollQuestion = message.media.poll.question;
        mediaDescriptions.push(`Poll: "${pollQuestion}"`);
        break;
        
      case 'MessageMediaWebPage':
        const webPage = message.media.webpage;
        if (webPage.title) {
          mediaDescriptions.push(`Webpage: "${webPage.title}"`);
        } else if (webPage.url) {
          mediaDescriptions.push(`Webpage: ${webPage.url}`);
        } else {
          mediaDescriptions.push("Webpage");
        }
        break;
        
      default:
        mediaDescriptions.push(`Media of type: ${mediaType}`);
    }
  }
  
  return mediaDescriptions;
}

/**
 * Main function to run the Telegram client
 */
async function main() {
  console.log('Starting Safe Telegram Auto Commenter...');
  console.log(`Rate limits: ${MAX_COMMENTS_PER_HOUR} comments per hour, ${MAX_COMMENTS_PER_DAY} comments per day`);

  // Load or create session
  let stringSession;
  if (fs.existsSync(SESSION_FILE)) {
    const sessionData = fs.readFileSync(SESSION_FILE, 'utf8');
    stringSession = new StringSession(sessionData);
    console.log('Loaded existing session');
  } else {
    stringSession = new StringSession('');
    console.log('Created new session');
  }

  // Create client
  const client = new TelegramClient(
    stringSession,
    apiId,
    apiHash,
    { connectionRetries: 5 }
  );

  // Connect and login
  await client.start({
    phoneNumber: async () => phoneNumber,
    password: async () => password, // Use saved password
    phoneCode: async () => await input.text('Please enter the code you received: '),
    onError: (err) => console.log(err),
  });

  // Save session string
  const sessionString = client.session.save();
  fs.writeFileSync(SESSION_FILE, sessionString);
  console.log('Session saved');

  // Get the list of channels from the environment variable
  const targetChannels = (process.env.TARGET_CHANNEL || '').split(',').map(ch => ch.trim()).filter(ch => ch);
  
  if (targetChannels.length === 0) {
    console.error('TARGET_CHANNEL environment variable is empty or not set. Please add channel usernames, separated by commas.');
    process.exit(1);
  }

  console.log(`Found ${targetChannels.length} channel(s) to monitor in .env file.`);

  // Use a Map to store { channelId: discussionGroupId }
  const monitoredChannels = new Map();
  for (const channelUsername of targetChannels) {
    try {
      console.log(`Resolving channel: ${channelUsername}`);
      const entity = await client.getEntity(channelUsername);
      
      // Use GetFullChannel to reliably get linked chat information
      const fullChannel = await client.invoke(
        new Api.channels.GetFullChannel({ channel: entity })
      );

      const channelId = String(entity.id);
      // The discussion group ID is in fullChat.linkedChatId
      const discussionGroupId = fullChannel.fullChat.linkedChatId;

      if (discussionGroupId) {
        monitoredChannels.set(channelId, discussionGroupId);
        console.log(`Successfully resolved channel '${entity.title}'. Will post comments to linked group ${discussionGroupId}.`);
      } else if (entity.megagroup) {
        // Fallback for older supergroups or edge cases
        monitoredChannels.set(channelId, channelId);
        console.log(`Successfully resolved group '${entity.title}'. Will post comments directly.`);
      } else {
        console.warn(`Channel '${entity.title}' does not appear to have a comment section. Skipping.`);
      }
    } catch (error) {
      console.error(`Could not resolve channel "${channelUsername}". Error: ${error.message}. Skipping.`);
    }
  }

  if (monitoredChannels.size === 0) {
    console.error("Fatal: Could not resolve any channels with linked discussion groups. Please check the usernames in your .env file. Exiting.");
    process.exit(1);
  }

  console.log(`Monitoring ${monitoredChannels.size} channels with discussion groups in total.`);
  console.log("Adding event handler for new messages...");

  // Add event handler for new messages
  client.addEventHandler(async (event) => {
    const message = event.message;
    if (!message || !message.peerId || !message.peerId.channelId) {
      return; // Not a channel message
    }

    const messageChannelIdStr = String(message.peerId.channelId);

    // Check if the message is from one of the monitored channels
    if (monitoredChannels.has(messageChannelIdStr)) {
      const discussionGroupId = monitoredChannels.get(messageChannelIdStr);
      
      console.log(`MATCH! New post in monitored channel ${message.peerId.channelId} at ${new Date().toISOString()}`);
      console.log(`Post content: ${message.text ? `"${message.text.substring(0, 70)}..."` : "[No Text]"}`);
      
      const mediaDescriptions = await processMedia(client, message);
      if (mediaDescriptions.length > 0) {
        console.log(`Post contains media: ${mediaDescriptions.join(', ')}`);
      }
      
      console.log("Checking rate limits...");
      await rateLimiter.handleComment();
      
      const comment = await generateComment(message, mediaDescriptions);
      console.log(`Generated comment: ${comment}`);
      
      try {
        // Reply to the post in the linked discussion group
        await client.sendMessage(discussionGroupId, {
          message: comment,
          replyTo: message.id
        });
        console.log(`Comment posted successfully to discussion group ${discussionGroupId}.`);
      } catch (error) {
        console.error('Error posting comment:', error);
        rateLimiter.hourHistory.pop();
        rateLimiter.dayHistory.pop();
      }
    }
  }, new NewMessage({}));

  console.log('Bot is running. Press Ctrl+C to stop.');
}

main().catch(console.error); 

File: safe_commenter.py
import os
import asyncio
import base64
from telethon import TelegramClient, events
from telethon.sessions import StringSession
from telethon.tl.types import MessageMediaPhoto, MessageMediaDocument
from telethon.tl.functions.channels import GetFullChannelRequest
from telethon.tl.functions.messages import GetDiscussionMessageRequest
from dotenv import load_dotenv
import aiohttp
from pathlib import Path
from datetime import datetime, timedelta
from io import BytesIO

# Загрузка переменных окружения
load_dotenv()

API_ID = int(os.getenv('TELEGRAM_API_ID'))
API_HASH = os.getenv('TELEGRAM_API_HASH')
PHONE = os.getenv('TELEGRAM_PHONE_NUMBER')
PASSWORD = os.getenv('TELEGRAM_PASSWORD')
TARGET_CHANNELS = [ch.strip() for ch in os.getenv('TARGET_CHANNEL', '').split(',') if ch.strip()]
OPENROUTER_API_KEY = os.getenv('OPENROUTER_API_KEY')
OPENROUTER_API_URL = os.getenv('OPENROUTER_API_URL', 'https://openrouter.ai/api/v1')
MODEL_NAME = os.getenv('MODEL_NAME')  # Должна быть vision-модель типа qwen/qwen-2-vl-7b-instruct

SESSION_FILE = 'telegram_session.session'
TEMP_DIR = Path('./temp')
TEMP_DIR.mkdir(exist_ok=True)

MAX_COMMENTS_PER_HOUR = 8
MAX_COMMENTS_PER_DAY = 50
MAX_REPLIES_PER_THREAD = 3  # Новое ограничение на ветку

class RateLimiter:
    def __init__(self, max_per_hour, max_per_day, verbose=True):
        self.max_per_hour = max_per_hour
        self.max_per_day = max_per_day
        self.verbose = verbose
        self.hour_history = []
        self.day_history = []

    async def handle_comment(self):
        now = datetime.now()
        self.hour_history = [t for t in self.hour_history if now - t <= timedelta(hours=1)]
        self.day_history = [t for t in self.day_history if now - t <= timedelta(days=1)]
        
        print("Waiting 15 seconds before replying...")
        await asyncio.sleep(15)

        if len(self.hour_history) >= self.max_per_hour or len(self.day_history) >= self.max_per_day:
            if self.verbose:
                print("Rate limit exceeded. Waiting before posting next comment...")
            await asyncio.sleep(60)
            await self.handle_comment()
        else:
            self.hour_history.append(now)
            self.day_history.append(now)
            if self.verbose:
                print(f"Allowed to comment. Hour: {len(self.hour_history)}/{self.max_per_hour}, Day: {len(self.day_history)}/{self.max_per_day}")

async def download_media(client, message):
    """Скачивает медиа в BytesIO для base64"""
    if not message.media:
        return None
    
    file_bytes = BytesIO()
    await client.download_media(message.media, file=file_bytes)
    file_bytes.seek(0)
    return file_bytes

def media_to_base64(file_bytes):
    """Конвертирует BytesIO в base64 data URL"""
    media_type = "image/jpeg"  # по умолчанию
    if file_bytes.getvalue().startswith(b'\xFF\xD8'):  # JPEG
        media_type = "image/jpeg"
    elif file_bytes.getvalue().startswith(b'\x89PNG'):  # PNG
        media_type = "image/png"
    
    base64_image = base64.b64encode(file_bytes.getvalue()).decode('utf-8')
    return f"data:{media_type};base64,{base64_image}"

async def generate_comment(content, images_base64=[], reply_context=""):
    """Генерирует ответ с vision поддержкой"""
    if not content.strip():
        content = "[Сообщение без текста]"
    
    headers = {
        "Content-Type": "application/json",
        "Authorization": f"Bearer {OPENROUTER_API_KEY}",
        "HTTP-Referer": "https://web.telegram.org/k/#@kishkeklak",  # Требуется OpenRouter
        "X-Title": "AitishnicaBot",  # Требуется OpenRouter
    }
    
    system_msg = (
        "Ты — подписчик личного Telegram-канала, русская айтишница 20 лет, красивая славянка. "
        "Отвечаешь на посты в комментариях: старайся часто использовать мат (блять, нахуй, пиздец, заебали и тд), "
        "жалуешься на мир/софт, но с технарским взглядом (сравнивай с математикой, теорией вероятностей, математической статистикой, кодом, алгоритмами, багами). "
        
        "ПРАВИЛА АНАЛИЗА ПОСТА:\n"
        "1. ЕСТЬ ФОТО? → ОСНОВНОЙ УПОР НА НЕЁ (70% ответа). Распознай текст на фото, людей, объекты. "
           "Если девушка/парень на фото — СДЕЛАЙ КОМПЛИМЕНТ + технарский комментарий.\n"
        "2. ТОЛЬКО ТЕКСТ? → распознай суть, дай ответ на пост.\n"
        "3. ФОТО+ТЕКСТ? → фото важнее текста.\n"
        
        "Стиль: коротко (2-4 предложения), разговорно ('короче', 'а чё'), предложения СВЯЗНЫЕ, "
        "одна ЧЕТКАЯ мысль. Будь саркастичной, умной, иногда милой."
    )

    # Строим messages с поддержкой vision
    messages = [{"role": "system", "content": system_msg}]
    
    user_content = []
    if reply_context:
        user_content.append({"type": "text", "text": f"Контекст (ответ на это): {reply_context}\n\n"})
    
    user_content.append({"type": "text", "text": f'Ответь на это сообщение как эта айтишница: "{content}"'})
    
    # Добавляем изображения
    for img_b64 in images_base64:
        user_content.append({
            "type": "image_url",
            "image_url": {"url": img_b64}
        })
    
    messages.append({
        "role": "user", 
        "content": user_content
    })

    data = {
        "model": MODEL_NAME,
        "messages": messages,
        "temperature": 0.8,
        "max_tokens": 500
    }

    async with aiohttp.ClientSession() as session:
        try:
            async with session.post(f"{OPENROUTER_API_URL}/chat/completions", 
                                  headers=headers, json=data) as resp:
                if resp.status == 200:
                    res = await resp.json()
                    return res['choices'][0]['message']['content']
                else:
                    print(f"OpenRouter API error: {resp.status} - {await resp.text()}")
                    return "Пхах, блять, API опять лег, короче типичный пиздец."
        except Exception as e:
            print(f"Error generating comment: {e}")
            return "Окак, серверы опять наебнулись, заебали."

async def main():
    print("Starting Vision Telegram Auto Commenter...")
    print(f"Rate limits: {MAX_COMMENTS_PER_HOUR}/hour, {MAX_COMMENTS_PER_DAY}/day, {MAX_REPLIES_PER_THREAD}/thread")

    if not all([PASSWORD, OPENROUTER_API_KEY, MODEL_NAME]):
        raise Exception("Missing required env vars")

    rate_limiter = RateLimiter(MAX_COMMENTS_PER_HOUR, MAX_COMMENTS_PER_DAY)
    
    # Трекинг веток бота: {discussion_msg_id: count_replies}
    bot_threads = {}
    session_str = Path(SESSION_FILE).read_text() if Path(SESSION_FILE).exists() else ''

    client = TelegramClient(StringSession(session_str), API_ID, API_HASH)
    await client.start(phone=PHONE, password=PASSWORD)
    
    Path(SESSION_FILE).write_text(client.session.save())
    print("Client started")

    if not TARGET_CHANNELS:
        print("No TARGET_CHANNEL in .env")
        return

    monitored_channels = {}
    for channel_username in TARGET_CHANNELS:
        entity = await client.get_entity(channel_username)
        full_channel = await client(GetFullChannelRequest(channel=entity))
        channel_id = str(entity.id)
        
        linked_chat_id = getattr(full_channel.full_chat, 'linked_chat_id', None)
        if linked_chat_id:
            discussion_chat = await client.get_entity(linked_chat_id)
            monitored_channels[channel_id] = discussion_chat
            print(f"Channel '{entity.title}' -> discussion {discussion_chat.id}")


    @client.on(events.NewMessage)
    async def handler(event):
        message = event.message

        # Игнорируем служебные и вообще всё без peer_id
        if not message.peer_id or not message or not getattr(message.peer_id, 'channel_id', None):
            return

        # ---------- 1. НОВЫЙ ПОСТ В КАНАЛЕ ----------
        if getattr(message, "post", False) and hasattr(message.peer_id, "channel_id"):
            message_channel_id = str(message.peer_id.channel_id)
            if message_channel_id not in monitored_channels:
                return

            discussion_entity = monitored_channels[message_channel_id]
            print(f"New channel post {message.id} in channel {message_channel_id}")

            await rate_limiter.handle_comment()

            images = []
            if message.media:
                img_bytes = await download_media(client, message)
                if img_bytes:
                    images.append(media_to_base64(img_bytes))
                    print("Downloaded image for vision (post)")

            comment = await generate_comment(message.text or "", images)
            print(f"Generated: {comment[:100]}...")

            # получаем id корневого сообщения в обсуждении
            discussion_message = await client(GetDiscussionMessageRequest(
                msg_id=message.id,
                peer=message.peer_id
            ))
            discussion_msg_id = discussion_message.messages[0].id

            sent_msg = await client.send_message(
                discussion_entity,  # entity, не id!
                comment,
                reply_to=discussion_msg_id
            )

            # регистрируем ветку (по корневому сообщению обсуждения)
            bot_threads[discussion_msg_id] = 1
            print(f"Bot reply #1 to post {message.id} -> discussion_msg {discussion_msg_id}, bot_msg {sent_msg.id}")
            return

        # ---------- 2. СООБЩЕНИЯ В ГРУППАХ ОБСУЖДЕНИЯ ----------
        # chat_id для сопоставления с discussion_entity.id
        peer = message.peer_id
        chat_id = getattr(peer, "channel_id", None) or getattr(peer, "chat_id", None)
        if not chat_id:
            return

        discussion_entity = None
        for channel_id, disc_entity in monitored_channels.items():
            if getattr(disc_entity, "id", None) == chat_id:
                discussion_entity = disc_entity
                break

        # если это не одна из наших групп обсуждения — выходим
        if not discussion_entity:
            return

        # это сообщение в нужной группе обсуждения
        print(f"Message in discussion group {chat_id}, msg_id={message.id}")

        # ---------- 2.1. Проверяем, это ли ответ в ветке бота ----------
        if (message.reply_to and
            getattr(message.reply_to, "reply_to_top_id", None) in bot_threads):

            thread_root_id = message.reply_to.reply_to_top_id
            reply_count = bot_threads.get(thread_root_id, 0)

            if reply_count >= MAX_REPLIES_PER_THREAD:
                print(f"Thread {thread_root_id} limit reached "
                      f"({reply_count}/{MAX_REPLIES_PER_THREAD})")
                return

            print(f"REPLY to bot thread {thread_root_id} "
                  f"(#{reply_count + 1}/{MAX_REPLIES_PER_THREAD})")

            await rate_limiter.handle_comment()

            images = []
            if message.media:
                img_bytes = await download_media(client, message)
                if img_bytes:
                    images.append(media_to_base64(img_bytes))
                    print("Downloaded reply image for vision")

            comment = await generate_comment(
                message.text or "",
                images,
                reply_context="Это ответ на твое предыдущее сообщение в ветке"
            )

            sent_msg = await client.send_message(
                discussion_entity,         # entity группы обсуждений
                comment,
                reply_to=message.id        # отвечаем на сообщение пользователя
            )

            bot_threads[thread_root_id] = reply_count + 1
            print(f"Bot reply #{bot_threads[thread_root_id]} in thread {thread_root_id}, "
                  f"bot_msg {sent_msg.id}")
            return

        # ---------- 2.2. Не ответ на ветку бота ----------
        print(
            f"Ignored reply in {chat_id}: "
            f"reply_to_top={getattr(message.reply_to, 'reply_to_top_id', None) if message.reply_to else None}"
        )


    print("Bot running... Ctrl+C to stop")
    await client.run_until_disconnected()

if __name__ == '__main__':
    asyncio.run(main())

